src/lib/cache/helper.go

```diff
@@ -0,0 +1,88 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/goharbor/harbor/src/lib/log"
+)
+
+var (
+	onceMu = keyMutex{m: &sync.Map{}}
+)
+
+// Once retrieve the cached key value, build and cache the value when the cached key value not found
+func Once(c Cache, key string, value interface{}, builder func() (interface{}, error), expiration ...time.Duration) error {
+	err := c.Fetch(key, value)
+	// value found from the cache
+	if err == nil {
+		return nil
+	}
+	// internal error
+	if !errors.Is(err, ErrNotFound) {
+		return err
+	}
+
+	// lock the key in cache and try to build the value for the key
+	onceKey := fmt.Sprintf("%p:%s", c, key)
+	onceMu.Lock(onceKey)
+
+	defer onceMu.Unlock(onceKey)
+
+	// fetch again to avoid to build the value multi-times
+	err = c.Fetch(key, value)
+	if err == nil {
+		return nil
+	}
+	// internal error
+	if !errors.Is(err, ErrNotFound) {
+		return err
+	}
+
+	val, err := builder()
+	if err != nil {
+		return err
+	}
+
+	if err := c.Save(key, val, expiration...); err != nil {
+		log.Warningf("failed to save value to cache, error: %v", err)
+
+		// save the val to cache failed, copy it to the value directly
+		return simpleCopy(value, val)
```

Comment:
Here you imply the `builder` MUST return the same type of data as the input parm `value`, which may not be true in some caces.

Reply:
Yes, they must be the same type. 

I don't think we will fetch a value which type is A from the cache,  and compute and return a value of type B when the key is not found in the cache.

Reply:
So we should carefully document that the `value` in parm of `FetchOrSave` must be a pointer to the return value of `builder`?

Also see my comment in the simple copy, should we consider the case when the return value of `builder` is a pointer?  Shall we fail early if it's a pointer?